Site creation: domain suggestion screen

nheagy opened the pull request:
Fixes #8181

Step 4 in the new site creation flow.

To test:
- In the My Sites list, use the `+` to enter the new site creation flow
- procede to step 4, domain suggestions
- Ensure you see suggestions based on the Site Title entered
- Ensure searching for new domains works

Needs review: @ScoutHarris 

_Note:_ The API will be getting changed to support `only_wordpressdotcom` parameter to limit suggestions to only `wpcom` domains. Until that lands, all domains will be suggested.

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,241 @@
+import UIKit
```
This isn't for Signup, so this filename is misleading. Maybe `SiteCreationDomainSuggestionViewController` or just `DomainSuggestionViewController`?

nheagy replied:
100% agree with this. I created that file weeks ago and didn't think about it since 😂

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,241 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        view.backgroundColor = WPStyleGuide.greyLighten30()
```
Just wondering - would this be better as `WPStyleGuide.configureColors(for: view, andTableView: tableView)`?

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SiteCreation.storyboard:
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
This is throwing a warning: `Unsupported Configuration: Prototype table cells must have reuse identifiers`.

ScoutHarris replied:
The Domain Suggestion nav bar doesn't match the design:
- The Back button should be "Back".
- The title should be "Create New Site".

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsActivityTableViewCell.xib:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This is throwing a warning: `Ambiguous Layout: Vertical position is ambiguous for "Activity Spinner".`

ScoutHarris replied:
And this constraint error:
<img width="330" alt="screen shot 2017-12-19 at 3 11 44 pm" src="https://user-images.githubusercontent.com/1816888/34181297-f41e4696-e4ce-11e7-9946-5ffcabf68c7f.png">

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        let description = NSLocalizedString("Pick an available \"yourname.wordpress.com\" address to let people find you on the web.", comment: "Description of how to pick a domain name during the site creation process")
+        let cell = LoginSocialErrorCell(title: title, description: description)
+        cell.selectionStyle = .none
+        return cell
```
The text formatting of this cell doesn't match the prior steps. I assume this is left to the future task of merging those steps cells and LoginSocialErrorCell?

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SiteCreationDomainSearchTableViewCell.xib:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
The border at the top of the search bar doesn't go all the way across to the left:
<img width="536" alt="search_bar" src="https://user-images.githubusercontent.com/1816888/34179588-d7eb648c-e4c8-11e7-9a68-694362f9e9f0.png">

ScoutHarris replied:
The search bar doesn't have placeholder text. Per the design, it should be "Type to get more suggestions".

ScoutHarris replied:
This is a bit nitpicky, but the icon doesn't line up with the text in the header & search results like it does in the design.

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,244 @@
+        cell.textLabel?.text = suggestion
+        cell.textLabel?.textColor = WPStyleGuide.darkGrey()
+        return cell
+    }
```
The cell separators don't go all the way to the left:
<img width="527" alt="screen shot 2017-12-19 at 2 35 26 pm" src="https://user-images.githubusercontent.com/1816888/34179897-f13ab1e4-e4c9-11e7-9527-47bb47d608c9.png">

ScoutHarris replied:
Per the design, the table rows should have disclosure indicators.

ScoutHarris replied:
The design shows the results text as having different colors, with the appended domain(?) in a lighter grey.

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SiteCreation.storyboard:
```diff
@@ -1,11 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
I'm not sure where this is happening, so I'll just put it here. The top border on the table looks to be a bit below the actual table top (i.e there's a white line above the border):
<img width="538" alt="screen shot 2017-12-19 at 2 43 35 pm" src="https://user-images.githubusercontent.com/1816888/34180190-0969f9e0-e4cb-11e7-90de-f18ff81e839c.png">

nheagy replied:
No clue what's going on there…

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,244 @@
+import UIKit
```
Shouldn't the view be portrait only on the iPhone?

ScoutHarris commented:
The initial searching screen looks... odd.

<img width="526" alt="searching" src="https://user-images.githubusercontent.com/1816888/34180466-0427362c-e4cc-11e7-84dc-fe16d01820da.png">

Is there a way to make this look better? Maybe show the activity indicator in the center of the view, hiding the table until it's loaded? Maybe @folletto has some thoughts...

ScoutHarris commented:
If you search with an empty search field, this happens (and the "loading" row never goes away):
<img width="448" alt="screen shot 2017-12-19 at 2 54 00 pm" src="https://user-images.githubusercontent.com/1816888/34180579-7f6bd7a2-e4cc-11e7-8bd2-2fb67c9fd007.png">

Also, if you tap on that "loading" row, it either does have a value (the confirmation alert is shown) or the app crashes. :)

ScoutHarris commented:
I found it odd that if you enter text and re-search, the original results that were based on the Site Title are retained and shown in the list. It seems like the results should should only contain matches to the search string.

ScoutHarris commented:
I tried to dismiss the keyboard, but there doesn't seem to be a way to do so. There is no tap off, and even Return on the keyboard doesn't do it.

ScoutHarris commented:
This is more of an organizational thing, but other Site Creation xibs and associated classes are in the NUX > Views directory. These are in the NUX > Controllers directory. To be consistent, can we agree on where they should be?

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SignupDomainSuggestionViewController.swift:
```diff
@@ -0,0 +1,244 @@
+
+extension SignupDomainSuggestionViewController: LoginWithLogoAndHelpViewController {
+    func handleHelpButtonTapped(_ sender: AnyObject) {
+        displaySupportViewController(sourceTag: .wpComLogin)
```
The other steps in the Site Creation process have new a sourceTag specific to the view. Can we do this here? Plus, this isn't Login. :)

folletto commented:
> I found it odd that if you enter text and re-search, the original results that were based on the Site Title are retained and shown in the list. It seems like the results should should only contain matches to the search string.

Yes. We have two possible approach there:

1. Show a loader in the search bar, in which case we keep the list below until some data is returned (this is probably a novel approach in the app tho)
2. Clear the list and show the spinner until we have new data.

I also agree it looks odd, so regarding `UIActivityIndicatorView`:

1. Not show the borders
2. Show the spinner, but half size of that (or snapped to the next small size we have already so it's consistent) — I guess `24px` should do?
3. Spinner color should be the dark (again, whatever we use elsewhere) — I guess our app `gray` color instead of Apple's gray?

nheagy commented:
I intentionally made the change to the site title suggestions around. When I was playing with it before that, it felt like my "good suggestions" based on the site title were lost, and that users would want to tap the back button just to get the site title-based suggestions to return. I can definitely turn that off so you can see what you think, @folletto @ScoutHarris.

folletto commented:
> that users would want to tap the back button just to get the site title-based suggestions to return.

This is an interesting consideration.

Do you feel users would do that more than just clearing with the `(×)` the search field?
Btw, can we maybe cache the initial "empty" value so when clearing it just instantly go back to it?

nheagy commented:
> can we maybe cache the initial "empty" value so when clearing it just instantly go back to it?

That sounded good to me so I started work on it 😀

Thanks @folletto & @ScoutHarris!

nheagy commented:
@ScoutHarris I think this one is ready for your re-review.

I merged in your changes to the `SiteCreation.storyboard` it's probably worth a check that all your recent changes are still there 😉 

Thanks!

ScoutHarris commented:
Hey @folletto , @nheagy - something I just noticed. 

The header text says to select a wordpress.com address:
<img width="451" alt="step_descr" src="https://user-images.githubusercontent.com/1816888/34307331-dc2aac40-e704-11e7-8cb5-e68eeaf24656.png">

However, the results show _all_ options:
<img width="537" alt="results" src="https://user-images.githubusercontent.com/1816888/34307355-f735bd54-e704-11e7-88ab-18f6acd7aa98.png">

That feels like it would be confusing. Maybe we can change the text to be more accurate/less specific?

ScoutHarris commented on WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsViewController.swift:
```diff
@@ -0,0 +1,253 @@
+import UIKit
```
The view rotates to landscape on iPhone. Maybe we should make it portrait only to be consistent?

ScoutHarris commented:
@nheagy - other than my questions about the suggestions and rotation, it looks good. The view looks much much better! 👍

nheagy commented:
> Unless I misunderstood what this does "only_wordpressdotcom": true, it looks like it's not working properly, and the suggestions should have only wordpress.com domains.

Ya, we're waiting for the patch to the endpoint. Once it's released, it will only show `wpcom` domains. I updated the description to mention that 👍 

> The view rotates to landscape on iPhone. Maybe we should make it portrait only to be consistent?

Debated updating the navigation controller to do this, but decided to leave that for now. Just added a `supportedInterfaceOrientations` implementation instead.

Should be good for another look @ScoutHarris!

ScoutHarris commented:
> we're waiting for the patch to the endpoint

Sorry, I must have missed that. Thanks for adding a note!

ScoutHarris commented:
Looks good @nheagy !! :shipit:

nheagy commented:
Thanks @ScoutHarris 🎉
